### ISSUE TITLE
[Translation] Inconsistent tracked catalogue resources, depending on cache

### ISSUE THREAD

**Issue by apfelbox** [symfony/symfony#36161]
**Symfony version(s) affected**: 5.0.5

**Description**  
Can't get the tracked resources of translator messages catalog in cached container.

**How to reproduce**  
In any project, just add some translations and dump the catalogue for this locale using this code:

```php
public function someControllerAction (TranslatorInterface $translator)
{
    dd($translator->getCatalogue("de"));
}
```

The `resources` property of the catalogue will only be filled if you remove the cache directory before (`rm -rf var/cache/*`), but will be empty if called on a filled cache.

This makes things building on top of translation resources quite fragile, as the tracked resources can be wrong (and therefore it might use a stale config cache).

Empty cache:
![2020-03-21 at 19 30](https://user-images.githubusercontent.com/1032411/77233853-a2e8ce80-6baa-11ea-9cd9-3ea3f044cd8b.png)

Filled cache:
![2020-03-21 at 19 19](https://user-images.githubusercontent.com/1032411/77233855-a5e3bf00-6baa-11ea-8ac4-ad3a9b31b401.png)

If you wonder about the 3 additional entries in `messages`, the empty entries are not removed when it is not loaded from the cache:
![2020-03-21 at 19 31](https://user-images.githubusercontent.com/1032411/77233865-b2681780-6baa-11ea-9392-fb4a68b716bd.png)

**Possible Solution**  
Always populate the tracked resources in a message catalogue. The resources seem to be available in the translator itself.

**Additional context**

**Comment by fancyweb** [comment]
Status: reviewed

AFAIK, that is normal. Resources are lost when the catalogue is dumped. It should not be hard to preserve them but is this a bug fix or a feature request? 🤔

**Comment by nicolas-grekas** [comment]
Agreed!

**Comment by apfelbox** [comment]
Ehm @nicolas-grekas, even if this would be a feature request, is there another way to send it? Or why did you just close this issue?

**Comment by nicolas-grekas** [comment]
I closed because the reported behavior has been analyzed and reported as correct.
If this needs to be turned into a feature request, this would need a description of the use case before I suppose.

**Comment by apfelbox** [comment]
The use case should be pretty clear from the initial description, but I can reiterate it:

I want to build something that builds upon the translation catalogue, but what is lazily initialized. This is currently not really possible, as I have the alternative of either:

• don't cache it, which can make it really slow
• use the `ConfigCache`, but that only "randomly" works if the cache is created in the "correct" request.


So basically building anything that takes translations and does something with them (for example dynamically dumping a subset of them for consumption in other tools) is not possible with caching right now.

**Comment by nicolas-grekas** [comment]
Apparently, resources are dumped next to the cached catalogues as `.meta` files?
This means we could imagine a `CachedMessageCatalogue` class that overrides the `getResources()` method to lazily load this file on demande?
Up to give it a try?

**Comment by apfelbox** [comment]
I will take a look that week. Thanks for the tip where to start! 🚀

**Comment by apfelbox** [comment]
So, apparently there is no interest in improving this situation.

Here is the workaround, for everybody interested:

• if you are in a cache warmer phase (ie. the resources of the catalogue are not empty), initialize the cache.
• if you had a previous cache, you need to force remove it (as you don't know if it was initialized with the correct resources).
• To remove it, you need to generate the config cache, get the file path and manually remove it, as there doesn't seem to be another way to force an update.
• if you are not in a cache warmer phase (there is no direct way to check for that, so just check that ie. the translator resources of any catalogue are not empty (be sure to use a locale that actually has resources)).

Yeah, that works but is pretty hacky.
If somebody wants to improve it: the implementation in #36313 works, you just need to add tests. 👍

**Comment by apfelbox** [comment]
Keep in mind that it is impossible to build a lazily-initialized cache that builds on top of translations, as you need to generate your cache "at the correct time", which is the direct opposite of lazily initializing it.
